ocaml/alerts/certificate/certificate_check.ml

```diff
@@ -41,39 +41,29 @@ let generate_alert epoch (host, expiry) =
     (host, Some (message, alert))
 
 let execute rpc session existing_messages (host, alert) =
-  let host_uuid = XenAPI.Host.get_uuid rpc session host in
-  let messages_in_host =
-    List.filter
-      (fun (_, record) -> record.API.message_obj_uuid = host_uuid)
-      existing_messages
-  in
-  let is_outdated (ref, record) =
-    match alert with
-    | None ->
-        true
-    | Some (body, (name, priority)) ->
-        record.API.message_body <> body
-        || record.API.message_name <> name
-        || record.API.message_priority <> priority
-  in
   (* CA-342551: messages need to be deleted if the pending alert regard the
      same host and has newer, updated information.
      If the pending alert has the same metadata as the existing host message
      it must not be emmited and the existing message must be retained, this
      prevents changing the message UUID.
      In the case there are alerts regarding the host but no alert is pending
      they are not destroyed since no alert is automatically dismissed. *)
-  let outdated_messages = List.filter is_outdated messages_in_host in
-  let replacement_is_pending = List.exists is_outdated messages_in_host in
-  match (alert, replacement_is_pending) with
-  | Some (message, (alert, priority)), true ->
-      List.iter
-        (fun (self, _) -> XenAPI.Message.destroy rpc session self)
-        outdated_messages ;
-      ignore
-        (XenAPI.Message.create rpc session alert priority `Host host_uuid
-           message)
-  | _ ->
+  match alert with
+  | Some (message, (alert, priority)) ->
+      let host_uuid = XenAPI.Host.get_uuid rpc session host in
+      let (_ : [> `message] Client.Id.t API.Ref.t) =
+        XenAPI.Message.create rpc session alert priority `Host host_uuid message
```

Comment:
Won't this now always create a new message and thus change the message UUID contrary to what the comment states?

Reply:
One way to fix this would be perhaps to modify the previous commit to:
`match (alert, replacement_is_pending)` to `match (alert, replacement_is_pending || previous_messages = [])`. Also `replacement_is_pending` can be written as `outdated_messages <> []`.

Reply:
I'd put all the existing_messages into a Set, make any changes needed to the set (e.g. add new message), then diff the old and the new sets and figure out what messages you have to create or destroy (there are set subtraction functions). And for building the Set I'd ignore the message UUID (thus ensuring that the message's UUID won't change: it'll retain the UUID of the  first message of its kind, at least on OCaml >=4.03 according to the Set's docs).

Reply:
You're right this doesn't work to avoid replacing messages that are still valid. Using sets it's a bit difficult since there's a mismatch between existing messages and the message that may be emitted: I need the reference to be able to delete messages, but I can't have the reference of the newly created message until it's created. I'll think more about this.

Reply:
I've replaced the filter with a partition: If there are any elements that are still current no new alerts get emmitted.